src/scheme/HierarchicalLeastSquares.cpp

```diff
@@ -0,0 +1,312 @@
+/* Copyright 2022 CNRS-AIST JRL and CNRS-UM LIRMM */
+
+#include <tvm/scheme/HierarchicalLeastSquares.h>
+
+#include <tvm/LinearizedControlProblem.h>
+#include <tvm/constraint/abstract/LinearConstraint.h>
+#include <tvm/constraint/internal/LinearizedTaskConstraint.h>
+#include <tvm/solver/internal/SolverEvents.h>
+#include <tvm/utils/internal/map.h>
+
+#include <iostream>
+
+namespace tvm
+{
+
+namespace scheme
+{
+using namespace internal;
+using VET = requirements::ViolationEvaluationType;
+
+const internal::SchemeAbilities HierarchicalLeastSquares::abilities_ = {NoLimit,
+                                                                        {{GeneralLevel, {true, {VET::L2}}}},
+                                                                        false};
+
+bool HierarchicalLeastSquares::solve_(const LinearizedControlProblem & problem,
+                                      internal::ProblemComputationData * data) const
+{
+  if(problem.size() > static_cast<int>(problem.substitutions().substitutions().size()))
+  {
+    Memory * memory = static_cast<Memory *>(data);
+    return memory->solver->solve();
+  }
+  else
+  {
+    problem.variables().setZero();
+    return true;
+  }
+}
+
+void HierarchicalLeastSquares::updateComputationData_(const LinearizedControlProblem & problem,
+                                                      internal::ProblemComputationData * data) const
+{
+  using EventType = ProblemDefinitionEvent::Type;
+  solver::internal::SolverEvents se;
+
+  if(data->hasEvents())
+  {
+    Memory * memory = static_cast<Memory *>(data);
+
+    // If some events require to rebuild the data, we skip all other events.
+    for(const auto & e : memory->events())
+    {
+      if(e.type() == EventType::SubstitutionAddition || e.type() == EventType::SubstitutionRemoval)
+      {
+        resetComputationData(problem, memory);
+        return;
+      }
+    }
+    while(memory->hasEvents())
+    {
+      auto e = memory->popEvent();
+      switch(e.type())
+      {
+        case EventType::WeightChange: {
+          const auto & c = problem.constraintWithRequirements(e.typedEmitter<EventType::WeightChange>());
+          if(c.requirements->priorityLevel().value() == 0)
+            throw std::runtime_error(
+                "[HierarchicalLeastSquares::updateComputationData_] "
+                "HierarchicalLeastSquares does not allow to change the weight of a Task with priority 0.");
```

Comment:
You're right, this requirement is not necessary. I didn't fully pay attention to this part yet, because it is linked to the update. Yet I removed the non-necessary check (53e4d9c)